tests/gradients/core/test_pulse_generator_gradient.py

```diff
@@ -0,0 +1,1123 @@
+        def circuit(params):
+            qml.evolve(ham_single_q_const)(params, T)
+            return qml.probs(wires=0), qml.expval(qml.PauliZ(0))
+
+        params = [jnp.array(0.4)]
+        jac = jax.jacobian(circuit)(params)
+        p = params[0] * T
+        exp_jac = (jnp.array([-1, 1]) * jnp.sin(2 * p) * T, -2 * jnp.sin(2 * p) * T)
+        for j, e in zip(jac, exp_jac):
+            assert qml.math.allclose(j[0], e)
+
+    @pytest.mark.xfail
+    @pytest.mark.parametrize("time_interface", ["python", "numpy", "jax"])
+    def test_simple_qnode_jit(self, time_interface):
+        """Test that a simple qnode can be differentiated with pulse_generator."""
+        import jax
+        import jax.numpy as jnp
+
+        jax.config.update("jax_enable_x64", True)
+        dev = qml.device("default.qubit.jax", wires=1)
+        T = {"python": 0.2, "numpy": np.array(0.2), "jax": jnp.array(0.2)}[time_interface]
+        ham_single_q_const = qml.pulse.constant * qml.PauliY(0)
+
+        @qml.qnode(dev, interface="jax", diff_method=pulse_generator)
+        def circuit(params, T=None):
+            qml.evolve(ham_single_q_const)(params, T)
+            return qml.expval(qml.PauliZ(0))
+
+        params = [jnp.array(0.4)]
+        p = params[0] * T
+        exp_grad = -2 * jnp.sin(2 * p) * T
+        jit_grad = jax.jit(jax.grad(circuit))(params, T=T)
+        assert qml.math.isclose(jit_grad, exp_grad)
+
+    @pytest.mark.slow
+    def test_advanced_qnode(self):
+        """Test that an advanced qnode can be differentiated with pulse_generator."""
+        import jax
+        import jax.numpy as jnp
+
+        jax.config.update("jax_enable_x64", True)
+
+        params = [jnp.array(0.21), jnp.array(-0.171), jnp.array([0.05, 0.03, -0.1])]
+        dev = qml.device("default.qubit.jax", wires=2)
+        ham = (
+            qml.pulse.constant * qml.PauliX(0)
+            + (lambda p, t: jnp.sin(p * t)) * qml.PauliZ(0)
+            + jnp.polyval * (qml.PauliY(0) @ qml.PauliY(1))
+        )
+
+        def ansatz(params):
+            qml.evolve(ham)(params, 0.1)
+            return qml.expval(qml.PauliY(0) @ qml.PauliX(1))
+
+        qnode_pulse_grad = qml.QNode(
+            ansatz,
+            dev,
+            interface="jax",
+            diff_method=pulse_generator,
+        )
+        qnode_backprop = qml.QNode(ansatz, dev, interface="jax")
+
+        grad_pulse_grad = jax.grad(qnode_pulse_grad)(params)
+        assert dev.num_executions == 1 + 12  # one forward execution, dim(DLA)=6
+        grad_backprop = jax.grad(qnode_backprop)(params)
+
+        assert all(qml.math.allclose(r, e) for r, e in zip(grad_pulse_grad, grad_backprop))
+
+
+@pytest.mark.jax
+class TestPulseGeneratorDiff:
+    """Test that pulse_generator is differentiable."""
+
+    # pylint: disable=too-few-public-methods
+    @pytest.mark.slow
+    def test_jax(self):
+        """Test that pulse_generator is differentiable with JAX."""
```

Comment:
I'm not sure I understand what this is testing that is not tested already above. Can you elaborate?

Reply:
This computes the second order derivative :) `fun` computes the `pulse_generator` derivative, and then we do `jax.grad(fun)` towards the end of the test, checking that the full transform can be differentiated again.

Reply:
Oh wow thats cool! Can you clarify that in the doc string? Didnt get that at first sight 🙃